features/networkmonitor/impl/src/main/kotlin/io/element/android/features/networkmonitor/impl/DefaultNetworkMonitor.kt

```diff
@@ -5,97 +5,53 @@
-
-    override val connectivity: StateFlow<NetworkStatus> = callbackFlow {
-
-        /**
-         *  Calling connectivityManager methods synchronously from the callbacks is not safe.
-         *  So instead we just keep the count of active networks, ie. those checking the capability request.
-         *  Debounce the result to avoid quick offline<->online changes.
-         */
-        val callback = object : ConnectivityManager.NetworkCallback() {
-            private val activeNetworksCount = AtomicInteger(0)
-
-            override fun onLost(network: Network) {
-                if (activeNetworksCount.decrementAndGet() == 0) {
-                    trySendBlocking(NetworkStatus.Offline)
-                }
-            }
-
-            override fun onAvailable(network: Network) {
-                if (activeNetworksCount.incrementAndGet() > 0) {
-                    trySendBlocking(NetworkStatus.Online)
-                }
-            }
-        }
-        trySendBlocking(connectivityManager.activeNetworkStatus())
-        val request = NetworkRequest.Builder()
-            .addCapability(NetworkCapabilities.NET_CAPABILITY_VALIDATED)
-            .build()
-
-        connectivityManager.registerNetworkCallback(request, callback)
-        Timber.d("Subscribe")
-        awaitClose {
-            Timber.d("Unsubscribe")
-            connectivityManager.unregisterNetworkCallback(callback)
+    override val connectivity: StateFlow<NetworkStatus> = matrixClient
+        .syncService()
+        .syncState
+        .map { syncState ->
+            syncState.toNetworkStatus()
         }
-    }
-        .distinctUntilChanged()
-        .debounce(300)
         .onEach {
             Timber.d("NetworkStatus changed=$it")
         }
-        .stateIn(appCoroutineScope, SharingStarted.WhileSubscribed(), connectivityManager.activeNetworkStatus())
-
-    private fun ConnectivityManager.activeNetworkStatus(): NetworkStatus {
-        return activeNetwork?.let {
-            getNetworkCapabilities(it)?.getNetworkStatus()
-        } ?: NetworkStatus.Offline
-    }
-
-    private fun NetworkCapabilities.getNetworkStatus(): NetworkStatus {
-        val hasInternet = hasCapability(NetworkCapabilities.NET_CAPABILITY_VALIDATED)
-        return if (hasInternet) {
-            NetworkStatus.Online
-        } else {
-            NetworkStatus.Offline
-        }
+        .stateIn(
+            scope = appCoroutineScope, started = SharingStarted.WhileSubscribed(),
+            initialValue = matrixClient
+                .syncService()
+                .syncState
+                .value
+                .toNetworkStatus()
+        )
+
+    private fun SyncState.toNetworkStatus(): NetworkStatus = when (this) {
+        SyncState.Idle,
+        SyncState.Running,
+        SyncState.Error,
+        SyncState.Terminated -> NetworkStatus.Online
+        SyncState.Offline -> NetworkStatus.Offline
```

Comment:
Shouldn't only `Running` be `Online`?

Reply:
Since `NetworkStatus.Offline` will make the "no network" banner to be shown, I think it's better to be optimistic here.